### PR TITLE
Phase 1 Update create_bucket to take Bucket object within Storage Client

### PR DESCRIPTION
**Pull request by lbristol88**
Part of issue: #7762

@tswast 
@engelke

### CODE REVIEW

**Comment by lbristol88** [comment]
@engelke

File: storage/tests/unit/test_client.py [hunk 1]
```diff
@@ -409,7 +409,42 @@ def test_create_bucket_conflict(self):
         json_sent = http.request.call_args_list[0][1]["data"]
         self.assertEqual(sent, json.loads(json_sent))
 
-    def test_create_bucket_success(self):
+    def test_create_bucket_with_object_conflict(self):
+        from google.cloud.exceptions import Conflict
+        from google.cloud.storage.bucket import Bucket
+
+        PROJECT = "PROJECT"
```
Comment (tswast) [inline storage/tests/unit/test_client.py]: nit: since these are just local variables, use lower-case letters.

File: storage/tests/unit/test_client.py [hunk 2]
```diff
@@ -409,7 +409,42 @@ def test_create_bucket_conflict(self):
...
+        URI = "/".join(
+            [
+                client._connection.API_BASE_URL,
+                "storage",
+                client._connection.API_VERSION,
+                "b?project=%s" % (OTHER_PROJECT,),
+            ]
+        )
+        data = {"error": {"message": "Conflict"}}
+        sent = {"name": BUCKET_NAME}
```
Comment (tswast) [inline storage/tests/unit/test_client.py]: nit: I recommend moving this down near the assertion and calling it `json_expected` to align with `json_sent`.

File: storage/tests/unit/test_client.py [hunk 3]
```diff
@@ -441,6 +476,39 @@ def test_create_bucket_success(self):
...
+        bucket_obj = Bucket(client, BUCKET_NAME)
+        URI = "/".join(
+            [
+                client._connection.API_BASE_URL,
+                "storage",
+                client._connection.API_VERSION,
+                "b?project=%s" % (PROJECT,),
+            ]
+        )
+        sent = {"name": BUCKET_NAME, "billing": {"requesterPays": True}}
```
Comment (tswast) [inline storage/tests/unit/test_client.py]: In addition to `requesterPays`, let's set some properties on the bucket object in this test. The properties I most want to see are

1. ~[location](https://googleapis.github.io/google-cloud-python/latest/storage/buckets.html#google.cloud.storage.bucket.Bucket.location) - Which region is this bucket belonging to?~ (actually, that's not settable yet. Let's add that to the list in the redesign doc if it's not there yet)
2. [`storage_class`](https://googleapis.github.io/google-cloud-python/latest/storage/buckets.html#google.cloud.storage.bucket.Bucket.storage_class) - What kind of bucket is this?
3. [`requester_pays`](https://googleapis.github.io/google-cloud-python/latest/storage/buckets.html#google.cloud.storage.bucket.Bucket.requester_pays) - Leave the argument to `create_bucket` unset and set it on the Bucket, instead. (We should deprecate `requester_pays` on `create_bucket`. Add that to the redesign list, too.)

File: storage/tests/unit/test_client.py (hunk 1 above)
Comment (lbristol88) [reply storage/tests/unit/test_client.py]: Changed variable names to lower-case on all tests I'm working on.

File: storage/tests/unit/test_client.py (hunk 2 above)
Comment (lbristol88) [reply storage/tests/unit/test_client.py]: Renamed `sent` to `json_expected` on all relevant tests for consistency and moved down as directed.

File: storage/google/cloud/storage/bucket.py [hunk 4]
```diff
@@ -398,7 +398,7 @@ class Bucket(_PropertyMixin):
     https://cloud.google.com/storage/docs/storage-classes
     """
 
-    def __init__(self, client, name=None, user_project=None):
+    def __init__(self, client=None, name=None, user_project=None):
```
Comment (tswast) [inline storage/google/cloud/storage/bucket.py]: Let's make this update as a separate PR, as I think we'll want to show better error messages in the existing Bucket methods when `client` isn't supplied (for example, say what method to call instead on Client).

File: storage/google/cloud/storage/client.py [hunk 5]
```diff
@@ -202,7 +202,7 @@ def batch(self):
         """
         return Batch(client=self)
 
-    def get_bucket(self, bucket_name):
+    def get_bucket(self, bucket_or_name):
```
Comment (tswast) [inline storage/google/cloud/storage/client.py]: I'd prefer this to be a separate PR, too, but it's okay this once. In general a PR should correspond to 1 easily-described change (new feature / bug fix). Now this PR has 2 new features (but they are related, at least).

File: storage/tests/unit/test_client.py (hunk 3 above)
Comment (lbristol88) [reply storage/tests/unit/test_client.py]: Added storage_class property and adjusted requester_pays as directed. Included requester_pays and location comment in the redesign doc.

**Comment by lbristol88** [comment]
I also quickly added the update for the Bucket constructor to make the client optional since it was a very small change that didn't break anything. 

When working on the Bucket updates, changes will be in their own branch moving forward and be submitted in a separate pull request.

File: storage/google/cloud/storage/bucket.py (hunk 4 above)
Comment (lbristol88) [reply storage/google/cloud/storage/bucket.py]: Took this file out of review - please disregard, thanks!

File: storage/google/cloud/storage/client.py (hunk 5 above)
Comment (lbristol88) [reply storage/google/cloud/storage/client.py]: Took this file out of review - please disregard, thanks!

File: storage/google/cloud/storage/client.py [hunk 6]
```diff
@@ -265,8 +266,8 @@ def create_bucket(self, bucket_name, requester_pays=None, project=None):
         To set additional properties when creating a bucket, such as the
         bucket location, use :meth:`~.Bucket.create`.
 
-        :type bucket_name: str
-        :param bucket_name: The bucket name to create.
+        :type bucket_or_name: str or resource
```
Comment (tswast) [inline storage/google/cloud/storage/client.py]: Could you make the docstring for `create_bucket` follow the same pattern as we use in BigQuery. These `:type:` and `:param:` are Sphinx/RST-style, but we are migrating to Google-style.

Docs on Google-style are at https://sphinxcontrib-napoleon.readthedocs.io/en/latest/ with an example file at https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google

We had to do this for BigQuery too. Basically, as we update each method we migrate it to the Google-style.

Comment (lbristol88) [reply storage/google/cloud/storage/client.py]: Updated docstring to reflect Google-style.

**Comment by tswast** [comment]
Python 2.7 test failure is due to #7841 

```
____________ TestStorageNotificationCRUD.test_notification_explicit ____________
self = <tests.system.TestStorageNotificationCRUD testMethod=test_notification_explicit>
    def setUp(self):
        self.case_buckets_to_delete = []
>       self._initialize_topic()
tests/system.py:1181:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/system.py:1166: in _initialize_topic
    from google.cloud.pubsub_v1 import PublisherClient
../pubsub/google/cloud/pubsub_v1/__init__.py:19: in <module>
    from google.cloud.pubsub_v1 import subscriber
../pubsub/google/cloud/pubsub_v1/subscriber/__init__.py:17: in <module>
    from google.cloud.pubsub_v1.subscriber.client import Client
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    from __future__ import absolute_import
    import pkg_resources
    import os
    import grpc
    from google.api_core import grpc_helpers
    from google.oauth2 import service_account
    from google.cloud.pubsub_v1 import _gapic
    from google.cloud.pubsub_v1 import types
    from google.cloud.pubsub_v1.gapic import subscriber_client
    from google.cloud.pubsub_v1.gapic.transports import subscriber_grpc_transport
    from google.cloud.pubsub_v1.subscriber import futures
>   from google.cloud.pubsub_v1.subscriber._protocol import streaming_pull_manager
E     File "/tmpfs/src/github/google-cloud-python/pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py", line 198
E       print(f"{messages_percent}, {bytes_percent}")
E                                                  ^
E   SyntaxError: invalid syntax
```

**Comment by tseaver** [comment]
@tswast Hmm, I wasn't aware that this redesign had actually been approved:  in fact I asked about that in yesterdays storage sync call, and got a "no, still under consideration."

**Comment by tswast** [comment]
@tseaver I got the 👍 from @frankyn , who's the DPE that owns GCS.

We don't plan on making any breaking changes in the client with the redesign. The redesign is intended to ease the transition to gRPC.

**Comment by frankyn** [comment]
Sorry @tseaver, I did check it off with @tswast but didn't follow-up on the Github issue. 

I'm also not sure about the meeting you're referring. @crwilcox is this a new spin off from the original libraries weekly?
